src/Core/MapUtil/FHTileMap.cpp

```diff
@@ -690,6 +690,89 @@ void FHTileMap::correctTerrainTypes(Core::LibraryTerrainConstPtr dirtTerrain,
     });
 }
 
+void FHTileMap::clearRoads(std::vector<FHTown> m_towns)
+{
+    for (bool reverseOrder : {false, true}) {
+        for (int z = 0; z < m_depth; ++z) {
+            for (int y = (reverseOrder) ? m_height - 1 : 0; (reverseOrder) ? y >= 0 : y < m_height; (reverseOrder) ? --y : ++y) {
+                for (int x = (reverseOrder) ? m_width - 1 : 0; (reverseOrder) ? x >= 0 : x < m_width; (reverseOrder) ? --x : ++x) {
+                    const FHPos pos{ x, y, z };
+
+                    // process only road tiles
+                    if (!get(pos).isRoad())
+                        continue;
+                    // exclude town entrance tiles
+                    if (std::any_of(m_towns.begin(), m_towns.end(), [=](FHTown& town) { return ((town.m_pos.m_x - 2 == x) && (town.m_pos.m_y == y) && (town.m_pos.m_z == z)); }))
+                        continue;
+
+                    const Tile def;
+
+                    bool TL = getNeighbour(pos, -1, -1, def).isRoad();
+                    bool T  = getNeighbour(pos, +0, -1, def).isRoad();
+                    bool TR = getNeighbour(pos, +1, -1, def).isRoad();
+                    bool L  = getNeighbour(pos, -1, +0, def).isRoad();
+                    //bool X  = get(pos).isRoad();
+                    bool R  = getNeighbour(pos, +1, +0, def).isRoad();
+                    bool BL = getNeighbour(pos, -1, +1, def).isRoad();
+                    bool B  = getNeighbour(pos, +0, +1, def).isRoad();
+                    bool BR = getNeighbour(pos, +1, +1, def).isRoad();
+
+                    // road redundancy check: if road removal will not affect its orthogonal neighbouring roads mutual connection status
+                    // (which is decided whether they are connected by diagonal neighbours), such road is renundant. delete it!
+                    if ((T && L && TL || T && R && TR || B && L && BL || B && R && BR) && (!T || !L || TL) && (!T || !R || TR) && (!B || !L || BL) && (!B || !R || BR))
```

Comment:
that condition is very hard to understand. I'd prefer some visual pattern matching .

smth like I did for
`const std::vector<PatternMatcher> g_matchers{`

`
 return true
                   && t.TL.SD(s) && t.TR.SD(s)
                   && t.BL.N && t.BR.N;
 `